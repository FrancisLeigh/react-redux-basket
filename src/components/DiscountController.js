import React, { useState } from 'react'
import { connect } from 'react-redux'
import { applyDiscount, removeDiscount } from '../store/actions'

const DiscountController = ({ discount, applyDiscount, removeDiscount }) => {
  const [discountCode, setDiscount] = useState('')
  const handleRemoveDiscount = () => {
    setDiscount('')
    removeDiscount()
  }
  return (
    <div>
      <label className="flex flex--end flex--padd-x">
        Discount Code
        <input
        placeholder="XXXXXX"
        minLength="6" maxLength="6"
        onChange={e => setDiscount(e.target.value)}
        value={discountCode}
        className="input input--half-width"/>
      </label>

      <button
      disabled={discountCode.length !== 6}
      onClick={() => applyDiscount(discountCode)}>
        Apply Discount
      </button>

      {discount && <button
      onClick={() => handleRemoveDiscount()}>
        Remove Discount
      </button>}

      {discount && <span>{discount}% off</span>}
    </div>
  )
}

const mapDispatchToProps = dispatch => ({
  applyDiscount: discountCode => {
    return fetch(`http://localhost:3000/discounts/${discountCode}`)
    .then(res => {
      return dispatch(applyDiscount(Math.floor(Math.random() * 100) + 1))
    })
  },
  removeDiscount: () => dispatch(removeDiscount())
})

const mapStateToProps = ({ meta: { discount } }) => ({
  discount
})

export default connect(mapStateToProps, mapDispatchToProps)(DiscountController)
