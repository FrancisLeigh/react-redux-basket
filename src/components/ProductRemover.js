import React from 'react'
import { connect } from 'react-redux'
import { removeProduct } from '../store/actions'

const ProductRemove = ({id, removeProduct}) => (
  <button className="button button--clear" onClick={() => removeProduct(id)}>X</button>
)

const mapDispatchToProps = dispatch => ({
  removeProduct: id => dispatch(removeProduct(id))
})

export default connect(null, mapDispatchToProps)(ProductRemove)
