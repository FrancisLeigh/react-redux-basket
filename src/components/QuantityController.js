import React from 'react'
import { connect } from 'react-redux'
import { changeQuantity } from '../store/actions'

const QuantityController = ({ id, quantity, changeQuantity }) => (
  <div>
    <label className="accessible" htmlFor={id}>Quantity</label>
    <input id={id} className="quantity-ctrl" type="number" value={quantity} onChange={(e) => changeQuantity(id, e.target.value)} />
  </div>
)

const mapDispatchToProps = dispatch => ({
  changeQuantity: (id, quantity) => dispatch(changeQuantity(id, quantity))
})

export default connect(null, mapDispatchToProps)(QuantityController)
