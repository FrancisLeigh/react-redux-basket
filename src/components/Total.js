import React from 'react'
import { connect } from 'react-redux'
import { discountedTotalSelector } from '../store/selectors'

const Total = ({ total }) => (<React.Fragment>{total.toFixed(2)}</React.Fragment>)


const mapStatetoProps = state => ({
  total: discountedTotalSelector(state)
})

export default connect(mapStatetoProps)(Total)
