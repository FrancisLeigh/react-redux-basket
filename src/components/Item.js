import React from 'react'

import QuantityController from 'Components/QuantityController'
import ProductRemover from 'Components/ProductRemover'
const Item = ({ id, name, quantity, price }) => {
  return (
    <li className="basket__item flex--center flex--padd-x">
      <div className="basket__item__content">
        <span>{name}</span>
      </div>
      <div className="basket__item__content">
        <span className="basket__item__price">Â£{price.toFixed(2)}</span>
      </div>
      <div className="basket__item__content flex flex--row flex--center flex--end">
        <QuantityController id={id} quantity={quantity} name/>
        <ProductRemover id={id}/>
      </div>
    </li>
  )
}

export default Item
