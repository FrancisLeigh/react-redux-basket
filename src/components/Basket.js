import React from 'react'
import { connect } from 'react-redux'
import { clearQuantities } from '../store/actions'

import Item from './Item'
import BasketTotal from './Total'
const Basket = ({ products, clearQuantities }) => {
  return (
    <div className="basket">
      <div className="basket flex--wider basket flex--padd">
      <ul className="basket flex--wider basket flex--padd">
        {products.map((props) => (
            <Item key={`product-${props.id}`} {...props} />
          )
        )}
      </ul>
      <div className="basket__footer flex flex--row flex--center">
        <h1 className="flex flex--fit">
          Â£<BasketTotal />
        </h1>
        <div className="flex flex--fit flex--row">
          <button className="button" onClick={() => clearQuantities()}>Clear</button>
          <button className="button button--primary" disabled>Checkout ></button>
        </div>
      </div>
    </div>
  )
}

const mapStateToProps = state => ({
  products: state
})

const mapDispatchToProps = dispatch => ({
  clearQuantities: () => dispatch(clearQuantities())
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Basket)
