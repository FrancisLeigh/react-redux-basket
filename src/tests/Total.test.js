import React from 'react'
import { shallow } from 'enzyme'
import { Total } from '../components/Total'

const MockProps = { total: 100, discountedTotal: 50 }

describe('<Total />', () => {
  it('Should render total and discounted total', () => {
    const component = shallow(<Total  {...MockProps} />)

    expect(component).toMatchSnapshot()
  })
})
