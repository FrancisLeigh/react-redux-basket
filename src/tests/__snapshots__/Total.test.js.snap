// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Total /> Should render total and discounted total 1`] = `
<Fragment>
  <span>
    50.00
  </span>
  <span
    className="flex flex--padd-y"
    style={
      Object {
        "color": "red",
        "textDecoration": "line-through",
      }
    }
  >
    100.00
  </span>
</Fragment>
`;
